### PR TITLE
Add ability to build an RPM for RHEL/CentOS 7.

### PR DESCRIPTION
**Pull request by brilong**
Please review and provide feedback.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -87,24 +87,34 @@ cscope:
...
 
 ##
 # housekeeping
 ##
-clean: 
+clean:
 	rm -f cscope.out cscope.files
-	rm -f "$(DOCDIR)/joy.txt"
+	rm -f $(DOCDIR)/joy.txt
+	rm -f $(BINDIR)/joy $(BINDIR)/jfd-anon $(BINDIR)/joy-anon $(BINDIR)/str_match_test $(BINDIR)/unit_test
```
Comment (fliphil) [inline Makefile]: Don't need this line because bin/ is deleted in clobber target below.

File: build_rpm.sh [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+SHORT_COMMIT_ID=$(git show-ref --hash=7 refs/heads/$GITREF)
+TAR_FILE=joy-$SHORT_COMMIT_ID.tar
+TGZ_FILE=joy-$SHORT_COMMIT_ID.tar.gz
+
+git archive --format=tar --prefix="joy/" $COMMIT_ID > $TAR_FILE
+if [ "$?" != 0 ]; then
+    echo Failed to create git archive tar
+    exit 1
+fi
+#if [ "$GITREF" != "master" ]; then
```
Comment (fliphil) [inline build_rpm.sh]: Do you need this code that was commented out? Might have forgotten to remove it.

File: rpm/joy.spec [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+%global gittag0 refs/heads/master
+%global shortcommit0 %(c=%{commit0}; echo ${c:0:7})
+
+Summary: Capture and analyze network flow data for research, forensics and monitoring.
+Name: joy
+Version: 1.71
+Release: 0.%{shortcommit0}.1%{?dist}
+License: GPLv2
+Group: System Environment/Base
+#Source0:  https://github.com/cisco/%{name}/archive/%{commit0}.tar.gz#/%{name}-%{shortcommit0}.tar.gz
```
Comment (fliphil) [inline rpm/joy.spec]: Forgot to remove this comment?

File: install/install-sh [hunk 4]
```diff
@@ -69,23 +81,36 @@ while getopts "c:a:w:k:" arg; do
...
 	    KEYFILE=${OPTARG}
 	    ;;
+	p)
+	    echo "-p was triggered with option ${OPTARG}" 
+	    PREFIX=${OPTARG}
+	    ;;
+	r)
+	    echo "-r was triggered with option ${OPTARG}" 
+	    RPMBUILD=true
+        RPMBUILDROOT=${OPTARG}
```
Comment (fliphil) [inline install/install-sh]: Indent missing here.

File: install/install-sh [hunk 5]
```diff
@@ -42,17 +47,24 @@ CONFIGFILE=${CWD}/options.cfg
 ANONFILE=${APP_ROOT}/internal.net
 WATCHFILE=""
 KEYFILE=""
+NEWKEY=0
 TLS_FINGERPRINT_FILE=${APP_ROOT}/resources/tls_fingerprint.json
+PREFIX="/usr/local"
+RPMBUILDROOT=""
```
Comment (fliphil) [inline install/install-sh]: Could you rename this to more generic BUILD_ROOT?

File: install/install-sh [hunk 6]
```diff
@@ -126,27 +151,31 @@ if [ -f ${KEYFILE} ]; then
...
+	if [ -z "$RPMBUILD" ]; then
+    	echo "generating SSH public/private keypair ... "
+    	ssh-keygen -f ${KEYFILE} -P "" -t rsa -b 2048
+    	retval=$?
+    	if [ $retval -ne "0" ]; then
+			echo "error: could not generate SSH key ${KEYFILE}"
+			exit
+    	fi
+    	if [ ! -f ${KEYFILE} ]; then
+			echo "error: could not find keyfile ${KEYFILE}"
```
Comment (fliphil) [inline install/install-sh]: Could you please make sure indentation is 4 spaces? This appears to be 8 spaces

File: install/install-sh [hunk 7]
```diff
@@ -161,106 +190,108 @@ if [ "$sysname" == "Darwin" ]; then
...
     if [ $retval -ne "0" ]; then
-        echo "error: could not copy bin/joy to /usr/local/bin"
-        exit
+        echo "error: could not copy joy to ${PREFIX}/bin"
+        exit 1
     fi
 
     # Install query
-    cp ${APP_ROOT}/query.py /usr/local/bin/
+    cp ${APP_ROOT}/query.py ${PREFIX}/bin/
```
Comment (fliphil) [inline install/install-sh]: Can you please remove this whole block of code for query.py? It no longer exists.

File: install/install-sh [hunk 8]
```diff
@@ -161,106 +190,108 @@ if [ "$sysname" == "Darwin" ]; then
...
-    chmod 600 /usr/local/etc/joy/${KEYFILE}
+    if [ -z "$RPMBUILD" ]; then
+    	mv ${KEYFILE} ${PREFIX}/etc/joy/
+    	retval=$?
+    	if [ $retval -ne "0" ]; then
+        	echo "error: could not move ${KEYFILE} to ${PREFIX}/etc/joy"
+        	exit 1
+    	fi
+    	chmod 600 ${PREFIX}/etc/joy/${KEYFILE}
+	fi
```
Comment (fliphil) [inline install/install-sh]: Move indent to left.

File: install/install-sh [hunk 9]
```diff
@@ -269,21 +300,21 @@ if [ "$sysname" == "Darwin" ]; then
...
+        echo "error: could not copy joy.1 to ${PREFIX}/share/man/man1 (man page will be unavailable)"
+        exit 1
     fi
 
-    ls -l /usr/local/bin/joy
-    ls -l /usr/local/bin/query.py
-    ls -l /usr/local/etc/joy/*
-    ls -l /usr/local/share/man/man1/joy.1
+    ls -l ${PREFIX}/bin/joy
+    ls -l ${PREFIX}/bin/query.py
```
Comment (fliphil) [inline install/install-sh]: Please remove this line for query.py

File: install/install-sh [hunk 10]
```diff
@@ -307,231 +338,249 @@ elif [ "$sysname" == "Linux" ]; then
...
+        echo "error: could not copy joyq to ${PREFIX}/bin/"
+        exit 1
     fi
 
     # Install the sleuth python module (dependency of joyq)
-    pip install ${APP_ROOT}/sleuth/
+	if [ -z "$RPMBUILD" ]; then
+    	pip install ${APP_ROOT}/sleuth/
+	else
+    	pip install -t $PREFIX/lib/python ${APP_ROOT}/sleuth/
```
Comment (fliphil) [inline install/install-sh]: Why pip install under prefix here?

File: install/install-sh [hunk 11]
```diff
@@ -307,231 +338,249 @@ elif [ "$sysname" == "Linux" ]; then
...
-        if [ $retval -ne "0" ]; then
-            echo "error: could not copy to /etc/init.d"
-            exit
-        fi
-        chmod +x /etc/init.d/joy
-        retval=$?
-        if [ $retval -ne "0" ]; then
-            echo "error: could not chmod +x /etc/init.d/joy"
-            exit
+        create_directory $RPMBUILDROOT/etc/init.d
```
Comment (fliphil) [inline install/install-sh]: Please fix indentation for this whole block.

**Comment by brilong** [comment]
I have updated files per your feedback.  I'm not sure why it says Makefile conflicts.  Please review again and push if you feel it's ready.

**Comment by fliphil** [comment]
Thanks, I'll give it another look through!

**Comment by brilong** [comment]
Please don't push this just yet.  If you can review everything that would be great, but I may remove the rpm/joy.spec file and build_rpm.sh in favor of a new script using fpm to build both rpm and deb.

**Comment by fliphil** [comment]
Let me know whenever this is ready to go in.

**Comment by brilong** [comment]
I hope I didn't screw something up, but before committing my latest changes, I ran git pull --rebase to pull in all the latest origin/master changes.  Not sure why my commit included all those commit messages.  I tested creation of rpm and deb and it seems to be working well.

**Pull request merged**